rules/validator.go

```diff
@@ -0,0 +1,181 @@
+// THE SOFTWARE.
+
+package rules
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3metrics/filters"
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/policy"
+)
+
+// Validator validates a ruleset.
+type Validator interface {
+	// Validate validates a ruleset.
+	Validate(rs RuleSet) error
+}
+
+type validator struct {
+	opts ValidatorOptions
+}
+
+// NewValidator creates a new validator.
+func NewValidator(opts ValidatorOptions) Validator {
+	return &validator{opts: opts}
+}
+
+func (v *validator) Validate(rs RuleSet) error {
+	if err := v.validate(rs); err != nil {
+		return v.wrapError(err)
+	}
+	return nil
+}
+
+func (v *validator) validate(rs RuleSet) error {
+	// Only the latest (a.k.a. the first) view needs to be validated
+	// because that is the view that may be invalid due to latest update.
+	latest, err := rs.Latest()
+	if err != nil {
+		return fmt.Errorf("could not get the latest ruleset snapshot: %v", err)
+	}
+	if err := v.validateMappingRules(latest.MappingRules); err != nil {
+		return err
+	}
+	return v.validateRollupRules(latest.RollupRules)
+}
+
+func (v *validator) validateMappingRules(mrv map[string]*MappingRuleView) error {
+	namesSeen := make(map[string]struct{}, len(mrv))
+	for _, view := range mrv {
+		// Validate that no rules with the same name exist.
+		if _, exists := namesSeen[view.Name]; exists {
+			return NewRuleConflictError(fmt.Sprintf("mapping rule %s already exists", view.Name))
+		}
+		namesSeen[view.Name] = struct{}{}
+
+		// Validate that the filter is valid.
+		if err := v.validateFilters(view.Filters); err != nil {
+			return err
+		}
+
+		// Validate that the policies are valid.
+		t := v.opts.MetricTypeFn()(view.Filters)
+		for _, p := range view.Policies {
+			if err := v.validatePolicy(t, p); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (v *validator) validateRollupRules(rrv map[string]*RollupRuleView) error {
+	namesSeen := make(map[string]struct{}, len(rrv))
+	for _, view := range rrv {
+		// Validate that no rules with the same name exist.
+		if _, exists := namesSeen[view.Name]; exists {
+			return NewRuleConflictError(fmt.Sprintf("rollup rule %s already exists", view.Name))
+		}
+		namesSeen[view.Name] = struct{}{}
+
+		// Validate that the filter is valid.
+		if err := v.validateFilters(view.Filters); err != nil {
+			return err
+		}
+
+		// Validate that the policies are valid.
+		t := v.opts.MetricTypeFn()(view.Filters)
+		for _, target := range view.Targets {
+			for _, p := range target.Policies {
+				if err := v.validatePolicy(t, p); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// Validate that there are no conflictiing rollup targets.
+	return v.validateRollupTargets(rrv)
+}
+
+func (v *validator) validateRollupTargets(rrv map[string]*RollupRuleView) error {
+	seen := make([]RollupTarget, 0, len(rrv))
+	for _, view := range rrv {
+		for _, target := range view.Targets {
```

Comment:
Could you put this logic in the loop in validateRollupRules? You're already looping through all of the targets on line 107.

Reply:
Sure